packages/core/test/actions.test.ts

```diff
@@ -595,6 +595,127 @@ describe('entry/exit actions', () => {
       }, 50);
     });
   });
+
+  describe('when reaching a final state', () => {
+    // https://github.com/davidkpiano/xstate/issues/1109
+    it('exit actions should be called when invoked machine reaches its final state', (done) => {
+      let exitCalled = false;
+      let childExitCalled = false;
+      const childMachine = Machine({
+        exit: () => {
+          exitCalled = true;
+        },
+        initial: 'a',
+        states: {
+          a: {
+            type: 'final',
+            exit: () => {
+              childExitCalled = true;
+            }
+          }
+        }
+      });
+
+      const parentMachine = Machine({
+        initial: 'active',
+        states: {
+          active: {
+            invoke: {
+              src: childMachine,
+              onDone: 'finished'
+            }
+          },
+          finished: {
+            type: 'final'
+          }
+        }
+      });
+
+      interpret(parentMachine)
+        .onDone(() => {
+          expect(exitCalled).toBeTruthy();
+          expect(childExitCalled).toBeTruthy();
+          done();
+        })
+        .start();
+    });
+  });
+
+  describe('when stopped', () => {
+    it('exit actions should be called when stopping a machine', () => {
+      let exitCalled = false;
+      let childExitCalled = false;
+
+      const machine = Machine({
+        exit: () => {
+          exitCalled = true;
+        },
+        initial: 'a',
+        states: {
+          a: {
+            exit: () => {
+              childExitCalled = true;
+            }
+          }
+        }
+      });
+
+      const service = interpret(machine).start();
+      service.stop();
+
+      expect(exitCalled).toBeTruthy();
+      expect(childExitCalled).toBeTruthy();
+    });
+
+    it('should call each exit handler only once when the service gets stopped', () => {
+      const actual: string[] = [];
+      const machine = createMachine({
+        exit: () => actual.push('root'),
+        initial: 'a',
+        states: {
+          a: {
+            exit: () => actual.push('a'),
+            initial: 'a1',
+            states: {
+              a1: {
+                exit: () => actual.push('a1')
+              }
+            }
+          }
+        }
+      });
+
+      interpret(machine).start().stop();
+      expect(actual).toEqual(['root', 'a', 'a1']);
+    });
+
+    it('should call exit handlers in document order when the service gets stopped', () => {
```

Comment:
This is the only test added in this file. It's sort of basic and acts as a regression test here. Its implementation depends on the fact that transitions "mess up" the `state.configuration` order and this is an implementation detail - I think I'm going to add in a sec another, similar, test for parallel states, just in case (I know it will pass right now)

Reply:
Added those tests for parallel regions here:
https://github.com/statelyai/xstate/pull/2903/commits/6b5a57c4e9a05f53500f8857dafdd598b4b7953d
I've created 3 very similar tests but each one of them is "messing" up the `state.configuration` in a different order so I've decided to include all of those cases